fl4health/clients/moon_client.py

```diff
@@ -108,8 +99,77 @@ def update_before_train(self, current_server_round: int) -> None:
 
+        # Optimizing betas for the MK-MMD loss has not been done yet
+        self.betas_optimized = False
+
         return super().update_before_train(current_server_round)
 
+    def update_after_step(self, step: int) -> None:
+        if step > 0 and step % self.beta_update_interval == 0:
+            if self.mkmmd_loss_weights and len(self.old_models_list) > 0 and self.global_model:
+                # Get the feature distribution of the old, local and global models with evaluation mode
+                local_distribution, global_distribution, old_distribution = self.update_buffers(
+                    self.model, self.global_model, self.old_models_list[-1]
+                )
+                # Update betas for the MK-MMD loss
+                if self.mkmmd_loss_weights[0] != 0:
+                    self.mkmmd_loss_min.betas = self.mkmmd_loss_min.optimize_betas(
+                        X=local_distribution, Y=global_distribution, lambda_m=1e-5
+                    )
+                    log(INFO, f"Set optimized betas to minimize distance: {self.mkmmd_loss_min.betas.squeeze()}.")
+
+                if self.mkmmd_loss_weights[1] != 0:
+                    self.mkmmd_loss_max.betas = self.mkmmd_loss_max.optimize_betas(
+                        X=local_distribution,
+                        Y=old_distribution,
+                        lambda_m=1e-5,
+                    )
+                    log(INFO, f"Set optimized betas to maximize distance: {self.mkmmd_loss_max.betas.squeeze()}.")
+
+                # Betas have been optimized
+                self.betas_optimized = True
+
+        return super().update_after_step(step)
+
+    def update_buffers(
+        self, local_model: torch.nn.Module, global_model: torch.nn.Module, old_local_model: torch.nn.Module
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Update the feature buffer of the local and global features."""
+        assert isinstance(local_model, MoonModel)
+        assert isinstance(global_model, MoonModel)
+        assert isinstance(old_local_model, MoonModel)
+
+        local_buffer = []
+        global_buffer = []
+        old_buffer = []
+
+        init_state_local_model = local_model.training
+
+        # Set the models to evaluation mode
+        local_model.eval()
+        global_model.eval()
+        old_local_model.eval()
+        assert not local_model.training
+        assert not global_model.training
+        assert not old_local_model.training
+
+        with torch.no_grad():
+            for input, target in self.train_loader:
+                input, target = input.to(self.device), target.to(self.device)
+                _, local_features = local_model(input)
+                _, global_features = global_model(input)
+                _, old_local_features = old_local_model(input)
+
+                local_buffer.append(local_features["features"].reshape(len(local_features["features"]), -1))
```

Comment:
Similar to FENDA, it might be nice to have a helper function for flattening. Maybe we should pull this out into a util function or something to be share throughout?